### ISSUE TITLE
Migrate flutter/packages Cirrus tasks to LUCI

### ISSUE THREAD

**Issue by keyonghan** [flutter/flutter#114373]
This tracks the CI migration for plugins and packages repositories. More context: [go/flutter-plugins-packages-luci-migration](http://goto.google.com/flutter-plugins-packages-luci-migration)

**Plugins Mac-intel**:
- [x] darwin-lint_podspecs -> Mac_x64 lint_podspecs https://github.com/flutter/plugins/pull/6637/files
- [x] ios-platform_tests CHANNEL:master https://github.com/flutter/plugins/pull/6682
- [x] ios-platform_tests CHANNEL:stable https://github.com/flutter/plugins/pull/6682
- [x] macos-build_all_plugins CHANNEL:master https://github.com/flutter/plugins/pull/6671
- [x] macos-build_all_plugins CHANNEL:stable https://github.com/flutter/plugins/pull/6671

**Plugins Mac-arm:**
- [x] ios-build_all_plugins CHANNEL:master
- [x] ios-build_all_plugins CHANNEL:stable
- [x] macos-platform_tests CHANNEL:master
- [x] macos-platform_tests CHANNEL:stable

**Plugins Linux:**
- [x] Linux plugin_tools_tests
- [ ] repo_checks
  - [x] format
  - [x] license
  - [x] pubspec
  - [x] readme
  - [x] dependabot
  - [x] gradle
  - [x] publishable
  - [ ] version - Blocked on https://github.com/flutter/flutter/issues/130076
  - [x] federated_safety
- [x] dart_unit_tests CHANNEL:master
- [x] dart_unit_tests CHANNEL:stable
- [x] analyze CHANNEL:master
- [x] analyze CHANNEL:stable
- [x] Legacy-version-analyze CHANNEL:3.7.12
- [x] Legacy-version-analyze CHANNEL:3.3.10
- [x] downgraded_analyze
- [x] readme_excerpts - folded into `repo_checks`
- [x] Web-build_all_plugins CHANNEL:master
- [x] Web-build_all_plugins CHANNEL:stable
- [x] linux-platform_tests CHANNEL:master
- [x] linux-platform_tests CHANNEL:stable

**Plugins Linux (heavy):**

- [x] linux-custom_package_tests CHANNEL:master
- [x] linux-custom_package_tests CHANNEL:stable - Blocked on https://github.com/flutter/flutter/issues/130071
- [x] android-platform_tests CHANNEL:master
  - [x] Building and linting
  - [x] Unit tests
  - [x] FTL. Needs:
    - [x] https://github.com/flutter/flutter/issues/131075
    - Initial migration will not bring the Google Maps API key, since it turns out we don't absolutely need it for the current tests, so that will be added after the fact:
        - [ ] https://github.com/flutter/flutter/issues/131069
        - [ ] https://github.com/flutter/flutter/issues/131071
        - [ ] https://github.com/flutter/flutter/issues/131073
- [x] android-platform_tests CHANNEL:stable - See `master` above
- [x] android-build_all_plugins CHANNEL:master
- [x] android-build_all_plugins CHANNEL:stable
- [x] web-platform_tests CHANNEL:master
- [x] web-platform_tests CHANNEL:stable
- [x] web-benchmarks_test CHROMIUM_BUILD:950363 - [Merged into `custom_package_tests`](https://github.com/flutter/packages/pull/4429)
- [x] web-benchmarks_test CHROMIUM_BUILD:1097615 - Same

**Packages Mac-intel**:
- [x] macos-build_all_plugins CHANNEL:master
- [x] macos-build_all_plugins CHANNEL:stable

**Comment by godofredoc** [comment]
@drewroengoogle @ricardoamador @yusuf-goog it seems like all of you have been brainstorming about docker support for different reasons. You may want to have a quick sync.

**Comment by stuartmorgan** [comment]
Collecting some notes from my initial looks at converting things, and discussion in other issues:
- `repo_checks` has a lot of scripts that individually should be very easy to bring over, but migrating that task without something analogous to `always` would be a non-trivial loss of functionality (see discussion [here](https://github.com/flutter/flutter/issues/122629#issuecomment-1468878295)).
- `analyze` is probably trivial to migrate, with the caveat that:
  - `legacy_version_analyze` and `downgraded_analyze` are currently configured to run three more analyze steps in parallel with each other, but only after `analyze` has succeeded. I don't know if there's any way to express that in LUCI. (We could just always run all three, but that would be something of a waste since if `analyze` fails the others are almost guaranteed to fail the same way.)
- `web-build_all_packages`, `linux-build_all_packages`, and `android-build_all_packages` are probably easy to migrate, we just need the same standard dependencies that I'm sure flutter/flutter must already have for its e2e tests of those platforms.
- `linux-platform_tests` requires `xvfb` (or an actual X server). Last I heard that was still unsolved for flutter/flutter.
- `dart_unit_tests` is probably straightforward to migrate.
- `linux-custom_package_tests` needs Chrome. Currently it's being installed on the fly by a step in the script, but I assume there is an established way using LUCI dependencies (expressed in `.ci.yaml`) to do that instead. Absent a compelling reason, I would definitely prefer we manage Chrome in LUCI the same way in flutter/packages as we do in flutter/flutter (and I assume eng prod would prefer that as well), so this task would need some adjustment.
    - There may be repo tooling changes needed to find the Chrome path; I can help with that part if so.
    - Once this is ported, it will be a perfect test case for https://github.com/flutter/flutter/issues/120736 since I have code in one of the scripts run by that task that uses emulators locally, and is disabled in CI only because we don't have emulators; see discussion in https://github.com/flutter/flutter/issues/120736
- `android-platform_tests` needs some secrets; I'm not sure how that's managed in LUCI. It also needs some specific local dependencies for communicating with FTL (currently handled via Docker).
    - Once this is ported, we will very likely move some of its tests off of FTL and onto emulators.
- `web-platform_tests` and `web_benchmarks_test` have the same Chrome needs as `linux-custom_package_tests`. The benchmark tests also apparently need `libgbm-dev`.

As noted in https://github.com/flutter/flutter/issues/120732#issuecomment-1546173534 some the steps currently rely on Docker-installed dependencies, but in most—perhaps all—cases I would expect that we could move to a non-Docker solution.

**Comment by keyonghan** [comment]
Adding @yusuf-goog as an a co-assignee to this bug, as he is helping with remaining linux migration.

**Comment by godofredoc** [comment]
@keyonghan rather than continuing with the planning over a bug, would you mind meeting with @stuartmorgan to discuss the next steps? this will be helpful for setting the right expectations.

**Comment by keyonghan** [comment]
Based on sync up with @stuartmorgan , targeting to finish migration in Q3.

Existing development issues with cirrus
- Miss dashboard support
- Tree status is not reflecting the truth https://github.com/flutter/flutter/issues/128954
- Re-run of a cirrus test is not signaling well (it still shows failure though rerun is in progress)

Infra support is needed:

- [ ] `always` logic: https://github.com/flutter/flutter/issues/122629
- [ ] how chrome tests work
- [ ] FTL to emulator: web-platform_tests https://github.com/flutter/flutter/issues/120736
- [x] secretes management (LUCI already supports kms: https://flutter.googlesource.com/recipes/+/refs/heads/main/recipe_modules/kms/api.py)
  - Google maps API key
  - FTL GCP identity
- [x] xvfb support (seems already supported with latest ubuntu images: https://github.com/flutter/flutter/issues/90676)

**Comment by stuartmorgan** [comment]
Another area where I'll need infra support that I found while doing more migrations is to discuss how to best handle the N-1 and N-2 legacy tests. In Cirrus we use the same `CHANNEL` env variable that we use to control the `flutter` checkout to set a flag to the the script (that tells it to skip packages that require a version of Flutter newer than that). We also have a Dart version that we pass as a flag that we need to stay in sync. Issues I hit:
- How can I access the channel from `.ci.yaml` in the script invocation?
- How can we keep the Flutter version (which is in `.ci.yaml`) and the Dart version in the same place?
- Can we make `properties` changes inline to update the value of the N-1 and N-2 Flutter versions as a single PR that takes effect in presubmit, or will the LUCI version be much harder to roll forward at each stable release?

**Comment by stuartmorgan** [comment]
@yusuf-goog How do secrets work in LUCI? I.e., what would the LUCI equivalent of [this kind of setup](https://github.com/flutter/packages/blob/2a508cb412f5554866e8e8a43b02bd38e88d5ce5/.cirrus.yml#L117-L118) (which are gcloud credentials for FTL and a Google Maps API key for integration testing a maps app) be?

**Comment by yusuf-goog** [comment]
> @yusuf-goog How do secrets work in LUCI? I.e., what would the LUCI equivalent of [this kind of setup](https://github.com/flutter/packages/blob/2a508cb412f5554866e8e8a43b02bd38e88d5ce5/.cirrus.yml#L117-L118) (which are gcloud credentials for FTL and a Google Maps API key for integration testing a maps app) be?

@stuartmorgan  take a look at this: go/flutter-luci-recipes#secrets-support-in-recipes

**Comment by stuartmorgan** [comment]
I'm going to close this as fixed; there's only one specific, pre-submit-only check still running in Cirrus, and it has a tracking issue, so the umbrella issue is no longer needed.
